service/history/handler.go

```diff
@@ -247,9 +247,13 @@ func (h *Handler) RecordActivityTaskHeartbeat(ctx context.Context, request *hist
 	}
 	workflowID := taskToken.GetWorkflowId()
 
-	engine, err1 := h.controller.GetEngine(ctx, namespaceID, workflowID)
-	if err1 != nil {
-		return nil, h.convertError(err1)
+	shardContext, err := h.controller.GetShardByNamespaceWorkflow(ctx, namespaceID, workflowID)
+	if err != nil {
+		return nil, h.convertError(err)
+	}
+	engine, err := shardContext.GetEngineWithContext(ctx)
+	if err != nil {
+		return nil, h.convertError(err)
```

Comment:
Why doing this all over the places? This belong to GetEngine method.

Reply:
https://github.com/temporalio/temporal/pull/2745/files#diff-45c4598dbf388574168a77db0f3c6c2dc52aa3f5a34bd9edbb6e796b10b9de9fR292

feel free to ask me to create a new API, i.e. `GenerateVClock()`

Reply:
Can this be done in history engine?

Reply:
of course, but my long term goal is to make history engine go away:
* API logic should be moved out, e.g. one API per package
* @yycptt will move the per shard / engine task processing logic out